File: site/src/lib/crossref.ts
export async function enrichWithCrossref(work: any) {
  try {
    if (work.doi && work.url) return work;
    const q = encodeURIComponent(work.title || "");
    const r = await fetch(`https://api.crossref.org/works?rows=1&query.title=${q}`);
    if (!r.ok) return work;
    const data = await r.json();
    const item = data?.message?.items?.[0];
    if (!item) return work;
    if (!work.doi && item.DOI) {
      work.doi = item.DOI;
      work.url = `https://doi.org/${item.DOI}`;
    }
    if (!work.url && item.URL) work.url = item.URL;
    if (!work.year && item.issued?.['date-parts']?.[0]?.[0]) {
      work.year = item.issued['date-parts'][0][0];
    }
    return work;
  } catch {
    return work;
  }
}

File: app/know_service/ext/crossref.py
import os

try:
    import requests
except Exception:  # pragma: no cover
    requests = None

CROSSREF_BASE = os.getenv("CROSSREF_BASE", "https://api.crossref.org")


def crossref_search(name: str, concepts, window=None):
    try:
        if not requests:
            return []
        query = name
        if concepts:
            query += " " + " ".join(concepts)
        r = requests.get(f"{CROSSREF_BASE}/works", params={"query": query, "rows": 5}, timeout=5)
        if r.ok:
            data = r.json()
            out = []
            for item in data.get("message", {}).get("items", []):
                title = item.get("title", [""])[0]
                year = None
                parts = item.get("issued", {}).get("date-parts", [[None]])[0]
                if parts:
                    year = parts[0]
                doi = item.get("DOI")
                url = item.get("URL")
                out.append({"title": title, "year": year, "doi": doi, "url": url})
            return out
    except Exception:
        pass
    return []
